diff --git a/tests/unit/hashexpire.tcl b/tests/unit/hashexpire.tcl
index 456987284..03f67dc40 100644
--- a/tests/unit/hashexpire.tcl
+++ b/tests/unit/hashexpire.tcl
@@ -287,9 +287,9 @@ start_server {tags {"hashexpire"}} {
     test {HGETEX - verify no change when field does not exist} {
         r FLUSHALL
         r HSET myhash f1 v1
-        set mem_before [r MEMORY USAGE myhash]
+        set mem_before [memory_usage myhash]
         assert_equal {{}} [r HGETEX myhash EX 1 FIELDS 1 f2]
-        set memory_after [r MEMORY USAGE myhash]
+        set memory_after [memory_usage myhash]
         assert_equal $mem_before $memory_after
     }
 
@@ -1452,7 +1452,7 @@ start_server {tags {"hashexpire"}} {
             r HEXPIRE myhash{t} 200 FIELDS 1 f1
 
             # Verify initial TTL state
-            set mem_before [r MEMORY USAGE myhash{t}]
+            set mem_before [memory_usage myhash{t}]
             assert_equal "v1 v2" [r HMGET myhash{t} f1 f2]
             assert_morethan [r HTTL myhash{t} FIELDS 1 f1] 100
             assert_equal -1 [r HTTL myhash{t} FIELDS 1 f2]
@@ -1471,7 +1471,7 @@ start_server {tags {"hashexpire"}} {
             }
 
             # Verify field values and TTLs are preserved
-            set memory_after [r MEMORY USAGE $newhash]
+            set memory_after [memory_usage $newhash]
             assert_equal "v1 v2" [r HMGET $newhash f1 f2]
             assert_morethan [r HTTL $newhash FIELDS 1 f1] 100
             assert_equal -1 [r HTTL $newhash FIELDS 1 f2]
@@ -1500,7 +1500,7 @@ start_server {tags {"hashexpire"}} {
         r copy myhash{t} nwhash{t}
 
         # Verify initial TTL state
-        assert_equal [r MEMORY USAGE myhash{t}] [r MEMORY USAGE nwhash{t}]
+        assert_equal [memory_usage myhash{t}] [memory_usage nwhash{t}]
         assert_equal "v1 v3 v4" [r HMGET myhash{t} f1 f3 f4]
         assert_equal "v1 v3 v4" [r HMGET nwhash{t} f1 f3 f4]
         assert_equal [r HEXPIRETIME myhash{t} FIELDS 1 f1] [r HEXPIRETIME nwhash{t} FIELDS 1 f1]
